cora/signal/lss.py

```diff
@@ -1379,3 +1418,154 @@ def za_density_sph(
+        nfreq = len(freq)
+        redshift = units.nu21 / freq - 1
+        freqmap = np.zeros(nfreq, dtype=[("centre", np.float64), ("width", np.float64)])
+        freqmap["centre"][:] = freq[:]
+        freqmap["width"][:] = np.abs(np.diff(freq[:])[0])
+
+        # Find index of central channel
+        ref_chan = int(nfreq / 2.0)
+
+        # Compute voxel volume
+        omega = healpy.nside2pixarea(self.nside)  # Pixel area in sr
+        if self.use_freq_dependent_voxel_volume:
+            dV = differential_comoving_volume(redshift)
+            dz = lssutil.calculate_width(redshift)
+        else:
+            dV = differential_comoving_volume(redshift[ref_chan])
+            dz = redshift[ref_chan + 1] - redshift[ref_chan]
+        voxvol = dV * dz * omega
+
+        # Make new map container and distribute in pixel, to make it easier to
+        # fill map with freq- and pol-dependent values
+        m = Map(freq=freqmap, polarisation=self.full_pol, nside=self.nside)
+        m.redistribute("pixel")
+
+        # Set standard deviation of map values
+        if self.variance is not None:
+            scale = self.variance**0.5
+        else:
+            scale = self.P_SN**0.5
+            if self.use_freq_dependent_voxel_volume:
+                scale /= voxvol[:, np.newaxis] ** 0.5
+            else:
+                scale /= voxvol**0.5
+
+        # Fill local section of map with random values
+        m.map[:].local_array[:, 0, :] = self.rng.normal(
+            scale=scale, size=(nfreq, m.map[:].local_shape[-1])
+        )
+
+        # Save voxel volume(s) as attribute
+        m.attrs["voxvol_ref"] = voxvol
+        m.attrs["central_redshift"] = redshift[ref_chan]
+
+        # Distribute in frequency, because other tasks often expect this
+        m.redistribute("freq")
+
+        if self._count == self.niter:
+            raise pipeline.PipelineStopIteration
+
+        return m
+
+
+def differential_comoving_volume(z, cosmo=None):
+    """Differential comoving volume at redshift z.
+
+    Useful for calculating the effective comoving volume.
+    For example, allows for integration over a comoving volume that has a
+    sensitivity function that changes with redshift. The total comoving
+    volume is given by integrating ``differential_comoving_volume`` to
+    redshift ``z`` and multiplying by a solid angle.
+
+    This is taken from:
+    https://docs.astropy.org/en/stable/_modules/astropy/cosmology/flrw/base.html#FLRW.differential_comoving_volume
+
+    Parameters
+    ----------
+    z : float
+     redshift
+    cosmo: Cosmology object
+        Default is cora.util.Cosmology() default.
+
+    Returns
+    -------
+    dV : float
+      Differential comoving volume per redshift per steradian at each
+      input redshift. unit (Mpc/h)^3
+    """
+
+    if cosmo is None:
+        cosmo = get_cosmo()
+
+    H_z = cosmo.H(z) * (
+        cosmo._unit_distance / 1000.0
+    )  # H(z) in unit [(km . h) / (Mpc . sec)]
```

Comment:
```suggestion
    # H(z) in unit [(km . h) / (Mpc . sec)]
    H_z = cosmo.H(z) * cosmo._unit_distance / 1000.0
```
I think this looks tidier this way, but it's optional

Reply:
Done